temporal-sdk/src/main/java/io/temporal/worker/slotsupplier/SlotSupplier.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker.slotsupplier;
+
+import java.util.Optional;
+
+public interface SlotSupplier<SlotInfo> {
+  // TODO: Needs to be cancellable
+  SlotPermit reserveSlot(SlotReservationContext ctx) throws InterruptedException;
+
+  Optional<SlotPermit> tryReserveSlot(SlotReservationContext ctx);
+
+  void markSlotUsed(SlotInfo info, SlotPermit permit);
+
+  void releaseSlot(SlotReleaseReason reason, SlotPermit permit);
+
+  /**
+   * Because we currently use thread pools to execute tasks, there must be *some* defined
+   * upper-limit on the size of the thread pool for each kind of task.
+   *
+   * <p>This method may be deprecated and removed later if and when the SDK supports virtual
```

Comment:
Just to note, this wouldn't be deprecated/removed since virtual threads are only in Java 21 and we still support Java 8